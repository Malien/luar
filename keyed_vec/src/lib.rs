use std::{
    iter::Enumerate,
    marker::PhantomData,
    ops::{Index, IndexMut},
};

#[derive(Clone, PartialEq, Eq)]
pub struct KeyedVec<K, V> {
    vec: Vec<V>,
    _key: PhantomData<K>,
}

impl<K, V> KeyedVec<K, V> {
    pub fn new() -> Self {
        Self {
            vec: Vec::new(),
            _key: PhantomData,
        }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            vec: Vec::with_capacity(capacity),
            _key: PhantomData,
        }
    }

    pub fn from_vec(vec: Vec<V>) -> Self {
        Self {
            vec,
            _key: PhantomData,
        }
    }

    pub fn into_values(self) -> std::vec::IntoIter<V> {
        self.vec.into_iter()
    }

    pub fn iter(&self) -> Iter<'_, K, V> {
        Iter {
            vec: self.vec.iter().enumerate(),
            _key: PhantomData,
        }
    }

    pub fn keys(&self) -> KeysIter<K> {
        KeysIter {
            current: 0,
            max: self.len(),
            _key: PhantomData,
        }
    }

    pub fn len(&self) -> usize {
        self.vec.len()
    }

    pub fn reserve(&mut self, additional: usize) {
        self.vec.reserve(additional)
    }

    pub fn slice(&self) -> &[V] {
        &self.vec
    }

    pub fn is_empty(&self) -> bool {
        self.vec.is_empty()
    }
}

impl<K, V> KeyedVec<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    pub fn push(&mut self, value: V) -> K {
        let key = self.next_key();
        self.vec.push(value);
        key
    }

    pub fn next_key(&self) -> K {
        self.vec.len().try_into().unwrap()
    }

    pub fn key_range(&self) -> std::ops::Range<K> {
        let start = 0.try_into().unwrap();
        let end = self.vec.len().try_into().unwrap();
        start..end
    }
}

impl<K, V> KeyedVec<K, V>
where
    K: Into<usize>,
{
    pub fn get(&self, key: K) -> Option<&V> {
        self.vec.get(key.into())
    }

    pub fn get_mut(&mut self, key: K) -> Option<&mut V> {
        self.vec.get_mut(key.into())
    }
}

impl<K, V> KeyedVec<K, V>
where
    K: Into<usize>,
    V: Clone,
{
    pub fn accommodate_for_key(&mut self, key: K, value: V) {
        let amount_missing = (key.into() + 1).checked_sub(self.vec.len());
        match amount_missing {
            Some(0) | None => {}
            Some(amount_missing) => self
                .vec
                .extend(std::iter::repeat(value).take(amount_missing)),
        }
    }
}

impl<K, V> Default for KeyedVec<K, V> {
    fn default() -> Self {
        Self {
            vec: Default::default(),
            _key: Default::default(),
        }
    }
}

impl<K, V> std::fmt::Debug for KeyedVec<K, V>
where
    V: std::fmt::Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_list().entries(&self.vec).finish()
    }
}

impl<K, V> Index<K> for KeyedVec<K, V>
where
    K: Into<usize>,
{
    type Output = V;

    fn index(&self, index: K) -> &Self::Output {
        &self.vec[index.into()]
    }
}

impl<K, V> IndexMut<K> for KeyedVec<K, V>
where
    K: Into<usize>,
{
    fn index_mut(&mut self, index: K) -> &mut Self::Output {
        &mut self.vec[index.into()]
    }
}

pub struct Iter<'a, K, V> {
    vec: Enumerate<std::slice::Iter<'a, V>>,
    _key: PhantomData<K>,
}

impl<'a, K, V> IntoIterator for &'a KeyedVec<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, &'a V);

    type IntoIter = Iter<'a, K, V>;

    fn into_iter(self) -> Self::IntoIter {
        Iter {
            vec: self.vec.iter().enumerate(),
            _key: PhantomData,
        }
    }
}

impl<'a, K, V> Iterator for Iter<'a, K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, &'a V);

    fn next(&mut self) -> Option<Self::Item> {
        self.vec
            .next()
            .map(|(idx, value)| (idx.try_into().unwrap(), value))
    }
}

pub struct IterMut<'a, K, V> {
    vec: Enumerate<std::slice::IterMut<'a, V>>,
    _key: PhantomData<K>,
}

impl<'a, K, V> IntoIterator for &'a mut KeyedVec<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, &'a mut V);

    type IntoIter = IterMut<'a, K, V>;

    fn into_iter(self) -> Self::IntoIter {
        IterMut {
            vec: self.vec.iter_mut().enumerate(),
            _key: PhantomData,
        }
    }
}

impl<'a, K, V> Iterator for IterMut<'a, K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, &'a mut V);

    fn next(&mut self) -> Option<Self::Item> {
        self.vec
            .next()
            .map(|(idx, value)| (idx.try_into().unwrap(), value))
    }
}

pub struct IntoIter<K, V> {
    vec: Enumerate<std::vec::IntoIter<V>>,
    _key: PhantomData<K>,
}

impl<K, V> IntoIterator for KeyedVec<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, V);

    type IntoIter = IntoIter<K, V>;

    fn into_iter(self) -> Self::IntoIter {
        IntoIter {
            vec: self.vec.into_iter().enumerate(),
            _key: PhantomData,
        }
    }
}

impl<K, V> Iterator for IntoIter<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = (K, V);

    fn next(&mut self) -> Option<Self::Item> {
        self.vec
            .next()
            .map(|(idx, value)| (idx.try_into().unwrap(), value))
    }
}

pub struct KeysIter<K> {
    current: usize,
    max: usize,
    _key: PhantomData<K>,
}

impl<K> Iterator for KeysIter<K>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    type Item = K;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current == self.max {
            None
        } else {
            self.current += 1;
            Some(self.current.try_into().unwrap())
        }
    }
}

#[macro_export]
macro_rules! keyed_vec {
    ($($expr:expr),*$(,)?) => {
        $crate::KeyedVec::from_vec(vec![$($expr),*])
    };
}
