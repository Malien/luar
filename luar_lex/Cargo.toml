[package]
name = "luar_lex"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
thiserror = "1.0"
lazy_static = "1.4.0"
logos = "0.12"
regex = "1.4"
num = "0.4"
quickcheck = { version = "1.0", optional = true }

[dev-dependencies]
test_util = { path = "../test_util" }
quickcheck = "1.0"
quickcheck_macros = "1.0"
insta = "1.3"
indoc = "1.0"