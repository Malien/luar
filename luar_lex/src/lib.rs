#[cfg(test)]
#[macro_use(quickcheck)]
extern crate quickcheck_macros;

pub mod format;
mod ident;
mod number_literal;
mod string_literal;
mod token;

pub type DynTokens = Box<dyn Iterator<Item = Token>>;

pub trait ToTokenStream {
    type Tokens: IntoIterator<Item = Token>;

    fn to_tokens(self) -> Self::Tokens;
}

impl<T: IntoIterator<Item = Token>> ToTokenStream for T {
    type Tokens = Self;

    fn to_tokens(self) -> Self::Tokens {
        self
    }
}

pub use ident::Ident;
pub use number_literal::NumberLiteral;
pub use string_literal::StringLiteral;
pub use token::Token;
