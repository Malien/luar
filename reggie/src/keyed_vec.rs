use std::{
    marker::PhantomData,
    ops::{Index, IndexMut},
};

#[derive(Clone, PartialEq, Eq)]
pub struct KeyedVec<K, V> {
    vec: Vec<V>,
    _key: PhantomData<K>,
}

impl<K, V> KeyedVec<K, V> {
    pub fn new() -> Self {
        Self {
            vec: Vec::new(),
            _key: PhantomData,
        }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            vec: Vec::with_capacity(capacity),
            _key: PhantomData,
        }
    }

    pub fn from_vec(vec: Vec<V>) -> Self {
        Self {
            vec,
            _key: PhantomData,
        }
    }

    pub fn values(self) -> std::vec::IntoIter<V> {
        self.vec.into_iter()
    }

    pub fn len(&self) -> usize {
        self.vec.len()
    }

    pub fn reserve(&mut self, additional: usize) {
        self.vec.reserve(additional)
    }

    pub fn slice(&self) -> &[V] {
        &self.vec
    }
}

impl<K, V> KeyedVec<K, V>
where
    K: TryFrom<usize>,
    K::Error: std::fmt::Debug,
{
    pub fn push(&mut self, value: V) -> K {
        let key = self.next_key();
        self.vec.push(value);
        key
    }

    pub fn next_key(&self) -> K {
        self.vec.len().try_into().unwrap()
    }
}

impl<K, V> KeyedVec<K, V>
where
    K: Into<usize>,
    V: Clone,
{
    pub fn accommodate_for_key(&mut self, key: K, value: V) {
        let amount_missing = (key.into() + 1).checked_sub(self.vec.len());
        match amount_missing {
            Some(0) | None => {}
            Some(amount_missing) => {
                self.vec.extend(std::iter::repeat(value).take(amount_missing))
            }
        }
    }
}

impl<K, V> Default for KeyedVec<K, V> {
    fn default() -> Self {
        Self {
            vec: Default::default(),
            _key: Default::default(),
        }
    }
}

impl<K, V> std::fmt::Debug for KeyedVec<K, V> where V: std::fmt::Debug {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_list().entries(&self.vec).finish()
    }
}

impl<K, V> Index<K> for KeyedVec<K, V>
where
    K: Into<usize>,
{
    type Output = V;

    fn index(&self, index: K) -> &Self::Output {
        &self.vec[index.into()]
    }
}

impl<K, V> IndexMut<K> for KeyedVec<K, V>
where
    K: Into<usize>,
{
    fn index_mut(&mut self, index: K) -> &mut Self::Output {
        &mut self.vec[index.into()]
    }
}

macro_rules! keyed_vec {
    ($($expr:expr),*$(,)?) => {
        $crate::keyed_vec::KeyedVec::from_vec(vec![$($expr),*])
    };
}

pub(crate) use keyed_vec;
