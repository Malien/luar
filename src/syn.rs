use std::num::NonZeroUsize;

use crate::lex::{NumberLiteral, StringLiteral, Token};
use nom::{alt, map, verify, Err, IResult, Needed, Parser};
use thiserror::Error;

/// Syntax
///
/// module -> { statement | function }
/// block -> { stat sc } [ret sc]
/// sc -> ';'
/// stat -> varlist1 '=' explist1
/// varlist1 -> var { ',' var }
/// var -> name | var '[' exp1 ']' | var '.' name
/// stat -> 'while' exp1 'do' block 'end'
///       | 'repeat' block 'until' exp1
///       | 'if' exp1 'then' block { elseif } ['else' block] 'end'
///       | functioncall
///       | tableconstructor
///       | 'local' declist [init]
/// elseif -> 'elseif' exp1 'then' block
/// ret -> 'return' explist
/// declist -> name { , 'name' }
/// init -> '=' explist1
/// exp -> '(' exp ')'
///      | 'nil'
///      | number
///      | literal
///      | var
///      | exp operator exp
///      | unaryoperator exp
///      | tableconstructor
///      | functioncall
/// tableconstructor -> '@' '(' [exp1] ')' | '@' [name] fieldlist
/// fieldlist -> '{' [ffieldlist1] '}' | '[' [lfieldlist1] ']'
/// ffieldlist1 -> ffield { ',' ffield }
/// ffield -> name '=' exp
/// lfieldlist1 -> exp { ',' exp }
/// functioncall -> var '(' [explist1] ')'
/// explist1 -> { exp1 ',' } exp
/// function -> 'function' name '(' [parlist1] ')' block 'end'
/// parlist1 -> 'name' { ',' name }

#[derive(Error, Debug)]
enum ParsingError {
    #[error("Invalid token found (expected {expected:?}, found {found:?})")]
    ExpectationFailure { expected: Token, found: Token },
    #[error("Unexpected EOF found (expected {expected:?})")]
    EOF { expected: Token },
}

fn expectation_failure<T>(expected: Token, found: Token) -> ParsingResult<'static, T> {
    Err(Err::Error(ParsingError::ExpectationFailure {
        expected,
        found,
    }))
}

type ParsingResult<'a, T> = IResult<&'a [Token], T, ParsingError>;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum Operator {}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
enum TableConstructor {}

#[derive(Debug, PartialEq, Clone)]
enum Expression {
    Nil,
    String(StringLiteral),
    Number(NumberLiteral),
    Variable {
        name: String,
    },
    BinaryOperator {
        lhs: Box<Expression>,
        op: Operator,
        rhs: Box<Expression>,
    },
    UnaryOperator {
        op: Operator,
        exp: Box<Expression>,
    },
    TableConstructor(TableConstructor),
    FunctionCall {
        name: String,
        args: Vec<Expression>,
    },
}

struct TokenParser {
    token: Token,
}

impl<'a> Parser<&'a [Token], Token, ParsingError> for TokenParser {
    fn parse(&mut self, input: &'a [Token]) -> IResult<&'a [Token], Token, ParsingError> {
        let res = take_one(input).and_then(|(rest, token)| {
            if token == self.token {
                Ok((rest, token))
            } else {
                Err(Err::Error(ParsingError::ExpectationFailure {
                    expected: self.token.clone(),
                    found: token,
                }))
            }
        });
        res
    }
}

fn take_one(input: &[Token]) -> ParsingResult<Token> {
    if input.len() > 0 {
        let token = input[0].clone();
        let rest = &input[1..];
        Ok((rest, token))
    } else {
        Err(Err::Incomplete(Needed::new(1)))
    }
}

fn parse_expr(input: &[Token]) -> ParsingResult<Expression> {
    let (rest, token) = take_one(input)?;
    match token {
        Token::Nil => Ok((rest, Expression::Nil)),
        Token::Number(literal) => Ok((rest, Expression::Number(literal))),
        Token::String(literal) => Ok((rest, Expression::String(literal))),
        _ => expectation_failure(Token::Nil, token), // TODO: expected expression, not Nil
    }
}

#[cfg(test)]
mod tests {
    use super::{parse_expr, Expression, ParsingError};
    use crate::lex::{NumberLiteral, StringLiteral, Token};

    type ReturnType = Result<(), nom::Err<ParsingError>>;

    #[test]
    fn nill_expr() -> ReturnType {
        assert_eq!(Expression::Nil, parse_expr(&[Token::Nil])?.1);

        Ok(())
    }

    #[quickcheck]
    fn number_expr(literal: NumberLiteral) -> ReturnType {
        let expression = parse_expr(&[Token::Number(literal)])?.1;
        match (&literal, &expression) {
            (NumberLiteral(x), Expression::Number(NumberLiteral(y)))
                if f64::is_nan(*x) =>
            {
                assert!(f64::is_nan(*y))
            }
            _ => assert_eq!(Expression::Number(literal), expression),
        };

        Ok(())
    }

    #[quickcheck]
    fn string_expr(literal: StringLiteral) -> ReturnType {
        assert_eq!(Expression::String(literal.clone()), parse_expr(&[Token::String(literal)])?.1);

        Ok(())
    }

    // fn 
}

// fn token<'a>(token: Token, input: &'a[Token]) -> ParsingResult<'a, Token> {
//     Ok((input[0], &input[1..]))
// }

// fn expression(input: &[Token]) -> ParsingResult<Token> {
//     return alt!(
//         map!()
//     );
// }
