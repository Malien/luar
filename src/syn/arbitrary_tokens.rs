use crate::lex::Token;

#[derive(Debug, Clone)]
pub struct ArbitraryTokens<T> {
    pub tokens: Vec<Token>,
    pub expected: T,
}

impl<T> From<(Vec<Token>, T)> for ArbitraryTokens<T> {
    fn from((tokens, expected): (Vec<Token>, T)) -> Self {
        Self { tokens, expected }
    }
}
