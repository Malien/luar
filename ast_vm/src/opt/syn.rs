use luar_lex::{Ident, NumberLiteral, StringLiteral};
use luar_syn::{BinaryOperator, UnaryOperator};
use non_empty::NonEmptyVec;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct Module {
    pub chunks: Vec<Chunk>,
    pub ret: Option<Return>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Chunk {
    FnDecl(FunctionDeclaration),
    Statement(Statement),
}

#[derive(Debug, Clone, PartialEq)]
pub struct FunctionDeclaration {
    pub name: FunctionName,
    pub args: Vec<LocalValueID>,
    pub body: Block,
}

#[derive(Debug, Clone, PartialEq)]
pub enum FunctionName {
    Plain(Var),
    Method(Var, Ident),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Block {
    pub statements: Vec<Statement>,
    pub ret: Option<Return>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    Assignment(Assignment),
    LocalDeclaration(Declaration),
    While(WhileLoop),
    Repeat(RepeatLoop),
    If(Conditional),
    FunctionCall(FunctionCall),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Return(pub Vec<Expression>);

#[derive(Debug, Clone, PartialEq)]
pub struct Assignment {
    pub names: NonEmptyVec<Var>,
    pub values: NonEmptyVec<Expression>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Declaration {
    pub names: NonEmptyVec<LocalValueID>,
    pub initial_values: Vec<Expression>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct WhileLoop {
    pub condition: Expression,
    pub body: Block,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RepeatLoop {
    pub body: Block,
    pub condition: Expression,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Conditional {
    pub condition: Expression,
    pub body: Block,
    pub tail: ConditionalTail,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ConditionalTail {
    End,
    Else(Block),
    ElseIf(Box<Conditional>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum FunctionCall {
    Method {
        func: Var,
        method: Ident,
        args: FunctionCallArgs,
    },
    Function {
        func: Var,
        args: FunctionCallArgs,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum FunctionCallArgs {
    Table(TableConstructor),
    Arglist(Vec<Expression>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct TableConstructor {
    pub lfield: Vec<Expression>,
    pub ffield: Vec<(Ident, Expression)>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Nil,
    String(StringLiteral),
    Number(NumberLiteral),
    Variable(Var),
    BinaryOperator {
        lhs: Box<Expression>,
        op: BinaryOperator,
        rhs: Box<Expression>,
    },
    UnaryOperator {
        op: UnaryOperator,
        exp: Box<Expression>,
    },
    TableConstructor(TableConstructor),
    FunctionCall(FunctionCall),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct LocalValueID(pub u16);
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct GlobalValueID(pub u16);

impl TryFrom<usize> for GlobalValueID {
    type Error = <u16 as TryFrom<usize>>::Error;

    fn try_from(value: usize) -> Result<Self, Self::Error> {
        u16::try_from(value).map(Self)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ValueID {
    Local(LocalValueID),
    Global(GlobalValueID),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Var {
    Named(ValueID),
    PropertyAccess {
        from: Box<Var>,
        property: Ident,
    },
    MemberLookup {
        from: Box<Var>,
        value: Box<Expression>,
    },
}

