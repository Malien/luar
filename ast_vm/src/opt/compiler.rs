use std::collections::HashMap;

use keyed_vec::KeyedVec;
use luar_lex::Ident;

use crate::{
    lang::LuaValue,
    opt::syn::{Chunk, FunctionDeclaration, FunctionName},
};

use super::syn::{
    Assignment, Block, Conditional, ConditionalTail, Declaration, Expression, FunctionCall,
    FunctionCallArgs, GlobalValueID, LocalValueID, Module, RepeatLoop, Return, Statement,
    TableConstructor, ValueID, Var, WhileLoop,
};

#[derive(Debug, Clone, Default)]
pub struct GlobalValues {
    pub cells: KeyedVec<GlobalValueID, LuaValue>,
    pub mapping: HashMap<String, GlobalValueID>,
}

pub struct LocalValues<'a> {
    pub current: LocalValueID,
    pub mapping: HashMap<String, LocalValueID>,
    pub globals: &'a mut GlobalValues,
}

impl GlobalValues {
    pub fn id_for(&mut self, ident: impl AsRef<str>) -> GlobalValueID {
        self.mapping
            .entry(ident.as_ref().to_owned())
            .or_insert_with(|| self.cells.push(LuaValue::Nil))
            .clone()
    }
}

impl<'a> LocalValues<'a> {
    fn id_for(&mut self, ident: impl AsRef<str>) -> ValueID {
        if let Some(&id) = self.mapping.get(ident.as_ref()) {
            return ValueID::Local(id);
        }
        return ValueID::Global(self.globals.id_for(ident));
    }

    fn delcare(&mut self, ident: Ident) -> LocalValueID {
        let id = self.current;
        self.current.0 += 1;
        self.mapping.insert(ident.into(), id);
        id
    }

    fn new(globals: &'a mut GlobalValues) -> Self {
        Self {
            current: LocalValueID(0),
            mapping: HashMap::new(),
            globals,
        }
    }

    fn globals(&mut self) -> &mut GlobalValues {
        self.globals
    }
}

pub fn compile_module(module: luar_syn::Module) -> Module {
    let mut global_state = GlobalValues::default();
    let mut root_locals = LocalValues::new(&mut global_state);

    let chunks = module
        .chunks
        .into_iter()
        .map(|chunk| match chunk {
            luar_syn::Chunk::FnDecl(decl) => Chunk::FnDecl(compile_fn_body(&mut root_locals, decl)),
            luar_syn::Chunk::Statement(statement) => {
                Chunk::Statement(compile_statement(&mut root_locals, statement))
            }
        })
        .collect();
    let ret = module.ret.map(|ret| compile_ret(&mut root_locals, ret));

    Module { chunks, ret }
}

fn compile_fn_body(
    root_locals: &mut LocalValues,
    decl: luar_syn::FunctionDeclaration,
) -> FunctionDeclaration {
    let name = match decl.name {
        luar_syn::FunctionName::Plain(var) => FunctionName::Plain(compile_var(root_locals, var)),
        luar_syn::FunctionName::Method(var, ident) => {
            FunctionName::Method(compile_var(root_locals, var), ident.clone())
        }
    };
    let globals = root_locals.globals();
    let mut fn_locals = LocalValues::new(globals);
    let args = decl
        .args
        .iter()
        .map(|ident| fn_locals.delcare(ident.clone()))
        .collect();
    let body = compile_block(&mut fn_locals, decl.body);
    FunctionDeclaration { name, args, body }
}

fn compile_statement(locals: &mut LocalValues, statement: luar_syn::Statement) -> Statement {
    match statement {
        luar_syn::Statement::Assignment(assignment) => Statement::Assignment(Assignment {
            names: assignment.names.map(|var| compile_var(locals, var)),
            values: assignment.values.map(|expr| compile_expr(locals, expr)),
        }),
        luar_syn::Statement::LocalDeclaration(decl) => {
            let initial_values = decl
                .initial_values
                .into_iter()
                .map(|expr| compile_expr(locals, expr))
                .collect();
            let names = decl.names.map_ref(|ident| locals.delcare(ident.clone()));
            Statement::LocalDeclaration(Declaration {
                names,
                initial_values,
            })
        }
        luar_syn::Statement::While(while_loop) => Statement::While(WhileLoop {
            condition: compile_expr(locals, while_loop.condition),
            body: compile_block(locals, while_loop.body),
        }),
        luar_syn::Statement::Repeat(repeat_loop) => Statement::Repeat(RepeatLoop {
            body: compile_block(locals, repeat_loop.body),
            condition: compile_expr(locals, repeat_loop.condition),
        }),
        luar_syn::Statement::If(conditional) => Statement::If(compile_if(locals, conditional)),
        luar_syn::Statement::FunctionCall(fn_call) => {
            Statement::FunctionCall(compile_fn_call(locals, fn_call))
        }
    }
}

fn compile_var(locals: &mut LocalValues, var: luar_syn::Var) -> Var {
    match var {
        luar_syn::Var::Named(ident) => Var::Named(locals.id_for(ident)),
        luar_syn::Var::PropertyAccess { from, property } => Var::PropertyAccess {
            from: Box::new(compile_var(locals, *from)),
            property,
        },
        luar_syn::Var::MemberLookup { from, value } => Var::MemberLookup {
            from: Box::new(compile_var(locals, *from)),
            value: Box::new(compile_expr(locals, *value)),
        },
    }
}

fn compile_expr(locals: &mut LocalValues, expr: luar_syn::Expression) -> Expression {
    match expr {
        luar_syn::Expression::Nil => Expression::Nil,
        luar_syn::Expression::String(str) => Expression::String(str),
        luar_syn::Expression::Number(num) => Expression::Number(num),
        luar_syn::Expression::Variable(var) => Expression::Variable(compile_var(locals, var)),
        luar_syn::Expression::BinaryOperator { lhs, op, rhs } => Expression::BinaryOperator {
            lhs: Box::new(compile_expr(locals, *lhs)),
            op,
            rhs: Box::new(compile_expr(locals, *rhs)),
        },
        luar_syn::Expression::UnaryOperator { op, exp } => Expression::UnaryOperator {
            op,
            exp: Box::new(compile_expr(locals, *exp)),
        },
        luar_syn::Expression::TableConstructor(tbl) => {
            Expression::TableConstructor(compile_table_constructor(locals, tbl))
        }
        luar_syn::Expression::FunctionCall(fn_call) => {
            Expression::FunctionCall(compile_fn_call(locals, fn_call))
        }
    }
}

fn compile_block(locals: &mut LocalValues, block: luar_syn::Block) -> Block {
    Block {
        statements: block
            .statements
            .into_iter()
            .map(|stmt| compile_statement(locals, stmt))
            .collect(),
        ret: block.ret.map(|ret| compile_ret(locals, ret)),
    }
}

fn compile_ret(locals: &mut LocalValues, ret: luar_syn::Return) -> Return {
    Return(
        ret.0
            .into_iter()
            .map(|expr| compile_expr(locals, expr))
            .collect(),
    )
}

fn compile_if(locals: &mut LocalValues, conditional: luar_syn::Conditional) -> Conditional {
    Conditional {
        condition: compile_expr(locals, conditional.condition),
        body: compile_block(locals, conditional.body),
        tail: compile_if_tail(locals, conditional.tail),
    }
}

fn compile_if_tail(locals: &mut LocalValues, tail: luar_syn::ConditionalTail) -> ConditionalTail {
    match tail {
        luar_syn::ConditionalTail::End => ConditionalTail::End,
        luar_syn::ConditionalTail::Else(block) => {
            ConditionalTail::Else(compile_block(locals, block))
        }
        luar_syn::ConditionalTail::ElseIf(conditional) => {
            ConditionalTail::ElseIf(Box::new(compile_if(locals, *conditional)))
        }
    }
}

fn compile_fn_call(locals: &mut LocalValues, fn_call: luar_syn::FunctionCall) -> FunctionCall {
    match fn_call {
        luar_syn::FunctionCall::Method { func, method, args } => FunctionCall::Method {
            func: compile_var(locals, func),
            method,
            args: compile_fn_call_args(locals, args),
        },
        luar_syn::FunctionCall::Function { func, args } => FunctionCall::Function {
            func: compile_var(locals, func),
            args: compile_fn_call_args(locals, args),
        },
    }
}

fn compile_table_constructor(
    locals: &mut LocalValues,
    tbl: luar_syn::TableConstructor,
) -> TableConstructor {
    TableConstructor {
        lfield: tbl
            .lfield
            .into_iter()
            .map(|expr| compile_expr(locals, expr))
            .collect(),
        ffield: tbl
            .ffield
            .into_iter()
            .map(|(ident, expr)| (ident, compile_expr(locals, expr)))
            .collect(),
    }
}

fn compile_fn_call_args(
    locals: &mut LocalValues,
    args: luar_syn::FunctionCallArgs,
) -> FunctionCallArgs {
    match args {
        luar_syn::FunctionCallArgs::Table(tbl) => {
            FunctionCallArgs::Table(compile_table_constructor(locals, tbl))
        }
        luar_syn::FunctionCallArgs::Arglist(args) => FunctionCallArgs::Arglist(
            args.into_iter()
                .map(|expr| compile_expr(locals, expr))
                .collect(),
        ),
    }
}

#[cfg(test)]
mod test {
    #[test]
    fn removes_named_local_var_lookup() {
        let program = "
            local a = 1
            local b = 2
            local c = a + b
            return c
        ";
        let ast = luar_syn::lua_parser::module(program).unwrap();
        let module = super::compile_module(ast);

        insta::assert_debug_snapshot!(module);
    }

    #[test]
    fn removes_named_var_lookup() {
        let program = "
            global = 1
            local l0
            local l1, l2 = 2, { global, l0; bar = l0 }
            function l2:foo(a, b, c)
                if a then
                    return global + l1
                elseif b then
                    return l1 - a
                else
                    return c + 1
                end
            end
            print('hello', global, l0 - l1.bar)
            return global, l1, l2, l2:foo(3, l1, l2[global]), unspecified
        ";
        let ast = luar_syn::lua_parser::module(program).unwrap();
        let module = super::compile_module(ast);

        insta::assert_debug_snapshot!(module);
    }
}
