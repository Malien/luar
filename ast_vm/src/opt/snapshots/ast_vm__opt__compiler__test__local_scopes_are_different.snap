---
source: ast_vm/src/opt/compiler.rs
expression: module
---
Module {
    chunks: [
        Statement(
            If(
                Conditional {
                    condition: Number(
                        NumberLiteral(
                            1.0,
                        ),
                    ),
                    body: Block {
                        statements: [
                            LocalDeclaration(
                                Declaration {
                                    names: NonEmptyVec(
                                        [
                                            LocalValueID(
                                                0,
                                            ),
                                        ],
                                    ),
                                    initial_values: [
                                        Number(
                                            NumberLiteral(
                                                42.0,
                                            ),
                                        ),
                                    ],
                                },
                            ),
                        ],
                        ret: None,
                    },
                    tail: End,
                },
            ),
        ),
        Statement(
            If(
                Conditional {
                    condition: Number(
                        NumberLiteral(
                            1.0,
                        ),
                    ),
                    body: Block {
                        statements: [
                            FunctionCall(
                                Function {
                                    func: Named(
                                        Global(
                                            GlobalValueID(
                                                0,
                                            ),
                                        ),
                                    ),
                                    args: Arglist(
                                        [
                                            UnaryOperator {
                                                op: Not,
                                                exp: Variable(
                                                    Named(
                                                        Global(
                                                            GlobalValueID(
                                                                1,
                                                            ),
                                                        ),
                                                    ),
                                                ),
                                            },
                                        ],
                                    ),
                                },
                            ),
                        ],
                        ret: None,
                    },
                    tail: End,
                },
            ),
        ),
    ],
    ret: None,
    local_count: 1,
}
